### ISSUE TITLE
[Question] Example for independent nginx deployment

### ISSUE THREAD

**Issue by lonix1** [shaarli/Shaarli#1455]
I already run dockerised nginx. I want to run dockerised shaarli, and proxy requests from nginx to shaarli.

All examples I've seen are for nginx running inside the shaarli container (which as an aside, is non-standard and was confusing to debug... that said I suppose it's convenient for beginners).

So:
- Is there a "normal" / stand-alone shaarli image - just the app and php-fpm, nothing else?
- If not, is there a working example of proxying requests from an existing nginx to shaarli? All nginx examples are for the server inside the container, and rely on `fastcgi_pass`, whereas in the normal setup where the proxy is independent, we would use `proxy_pass`.

Any advice appreciated!

**Comment by lonix1** [comment]
Okay after some tinkering, below is my solution.

My shaarli and nginx containers are on a shared network, that's why I can use the `shaarli` address.

````nginx
server {

  listen              443 ssl http2;
  listen              [::]:443 ssl http2;
  server_name         foo.example.com;    # populate this
  include             foo.conf;           # populate this, if necessary

  access_log          /var/log/nginx/shaarli-access.log;
  error_log           /var/log/nginx/shaarli-error.log error;

  proxy_set_header    X-Forwarded-Proto $scheme;
  proxy_set_header    X-Forwarded-Host  $host;
  proxy_set_header    X-Forwarded-For   $proxy_add_x_forwarded_for;
  proxy_set_header    X-Real-IP         $remote_addr;

  location / {
    #proxy_redirect   off;                # unsure whether necessary, works without
    #proxy_set_header Host $host;         # unsure whether necessary, works without
    proxy_pass        http://shaarli:80;
  }

}
````

This should be added to the docs.

**Comment by nodiscc** [comment]
The nginx reverseproxy configuration should be fixed in https://github.com/shaarli/Shaarli/pull/1389 - https://shaarli.readthedocs.io/en/doc-rework-setup/Reverse-proxy/. @lonix1 Is something else missing?

> Is there a "normal" / stand-alone shaarli image - just the app and php-fpm, nothing else?

The shaarli image is just that, Alpine + nginx + php-fpm https://github.com/shaarli/Shaarli/blob/master/Dockerfile

**Comment by virtadpt** [comment]
I think I see what you're getting at.
* Shaarli is installed in a container that only has php-fpm running in it.
* php-fpm has a network socket exposed for nginx to relay requests to.
* nginx is running in another container that
  * Connects to the Shaarli container at its php-fpm network socket.
  * Listens on another IP address for HTTP requests.
  * Forwards those requests to the php-fpm/Shaarli container.
  * Forwards responses from the php-fpm/Shaarli container back to the user.
* nginx -> nginx does seem kind of baroque, and we know it causes problems already.

**Comment by lonix1** [comment]
@virtadpt That's a very detailed explanation of the issue... thanks and yes!

In my (and most) setups, different containers are used for different services. Example:
- nginx container
- foo stack: app container, database container, redis container
- bar stack: app container, database container, redis container, admin container
- shaarli stack: app container   <---
- etc.

All of them are on the same network as nginx, which serves them all. Of course this is just one network design, but the most common.

Bundling nginx into the container (as done by shaarli) is really helpful to new docker users - so I don't think it's a bad thing. *But there should also be a "normal" shaarli image that does not include nginx.*

To make this work:
- a new `Dockerfile` without `/.docker/services.d/nginx/`
- tag it as `shaarli/shaarli:1.2.3`
- tag original image as `shaarli/shaarli:1.2.3-nginx` or `shaarli/shaarli:1.2.3-bundle`
- users should use my `.conf` file for nginx, as I posted above, it works

**I hope someone reopens this, and considers it! Thanks!!**

**Comment by ArthurHoaro** [comment]
I agree with everything that has been said. Embedding nginx is usefull if you want the Docker image to work out of the box, but in a lot of cases it can be an unnecessary layer.

> tag original image as shaarli/shaarli:1.2.3-nginx or shaarli/shaarli:1.2.3-bundle

I would rather tag the new image with something like `shaarli/shaarli:1.2.3-fpm` to avoid breaking existing installations.
